src/Http/Middleware/CsrfProtectionMiddleware.php

```diff
@@ -0,0 +1,184 @@
+ * used together your forms will have CSRF tokens automatically added
+ * when `$this->Form->create(...)` is used in a view.
+ */
+class CsrfProtectionMiddleware
+{
+    /**
+     * Default config for the CSRF handling.
+     *
+     *  - cookieName = The name of the cookie to send.
+     *  - expiry = How long the CSRF token should last. Defaults to browser session.
+     *  - secure = Whether or not the cookie will be set with the Secure flag. Defaults to false.
+     *  - httpOnly = Whether or not the cookie will be set with the HttpOnly flag. Defaults to false.
+     *  - field = The form field to check. Changing this will also require configuring
+     *    FormHelper.
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'cookieName' => 'csrfToken',
+        'expiry' => 0,
+        'secure' => false,
+        'httpOnly' => false,
+        'field' => '_csrfToken',
+    ];
+
+    /**
+     * Configuration
+     *
+     * @var array
+     */
+    protected $_config = [];
+
+    /**
+     * Constructor
+     *
+     * @param array $config Config options
+     */
+    public function __construct(array $config = [])
+    {
+        $this->_config = $this->_defaultConfig + $config;
+    }
+
+    /**
+     * Serve assets if the path matches one.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @param callable $next Callback to invoke the next middleware.
+     * @return \Psr\Http\Message\ResponseInterface A response
+     */
+    public function __invoke(ServerRequestInterface &$request, ResponseInterface $response, $next)
+    {
+        $cookies = $request->getCookieParams();
+        $cookieData = null;
+        if (isset($cookies[$this->_config['cookieName']])) {
+            $cookieData = isset($cookies[$this->_config['cookieName']]);
+        }
+
+        if (!empty($cookieData)) {
+            $params = $request->getAttribute('params');
+            $params['_csrfToken'] = $cookieData;
+            $request = $request->withAttribute('params', $params);
+        }
+
+        $method = $request->getMethod();
+        if ($method === 'requested') {
+            return $next($request, $response);
+        }
+
+        if ($method === 'GET' && $cookieData === null) {
+            $this->_setCookie($request, $response);
+
+            return $next($request, $response);
+        }
+
+        $this->_validateAndUnsetTokenField($request);
+
+        return $next($request, $response);
+    }
+
+    /**
+     * Checks if the request is POST, PUT, DELETE or PATCH and validates the CSRF token
+     *
+     * @param \Cake\Http\ServerRequest $request The request object.
+     * @return void
+     */
+    protected function _validateAndUnsetTokenField(ServerRequest &$request)
+    {
+        if (in_array($request->getMethod(), ['PUT', 'POST', 'DELETE', 'PATCH']) || $request->getData()) {
```

Comment:
I am not saying it matters in this case. In many cases using empty()/isset() instead of === null or array_key_exists do not matter.

I am just saying this should be the default. We should always be strict with types UNLESS we have to be weak with it for whatever reason.